### PR TITLE
[libc] fix clang-tidy llvm-header-guard warnings

### PR DESCRIPTION
**Pull request by nickdesaulniers**
Towards the goal of getting `ninja libc-lint` back to green, fix the numerous
instances of:

    warning: header guard does not follow preferred style [llvm-header-guard]

This is because many of our header guards start with `__LLVM` rather than
`LLVM`.

To filter just these warnings:

    $ ninja -k2000 libc-lint 2>&1 | grep llvm-header-guard
    
To automatically apply fixits:

    $ find libc/src libc/include libc/test -name \*.h | \
        xargs -n1 -I {} clang-tidy {} -p build/compile_commands.json \
        -checks='-*,llvm-header-guard' --fix --quiet
    
Some manual cleanup is still necessary as headers that were missing header
guards outright will have them inserted before the license block (we prefer
them after).

### CODE REVIEW

File: libc/include/llvm-libc-types/ACTION.h [hunk 1]
```diff
@@ -6,9 +6,9 @@
 //
 //===----------------------------------------------------------------------===//
 
-#ifndef __LLVM_LIBC_TYPES_ACTION_H__
-#define __LLVM_LIBC_TYPES_ACTION_H__
+#ifndef LLVM_LIBC_TYPES_ACTION_H__
```
Comment (nickdesaulniers) [inline libc/include/llvm-libc-types/ACTION.h:9]: @michaelrj-google I feel like I should clean up these trailing `__` as well. Thoughts?

**Comment by frobtech** [comment]
Is there a reason you used perl-as-sed rather than `clang-tidy --fix`? I think that would fully canonicalize things wrt the trailing underscores too.

**Comment by michaelrj-google** [comment]
It seems like all of these are public headers in `include`. I think we had those extra underscores for macro namespacing purposes. I'm not sure if we actually want to remove those, it might be better to remove the check from the include folder.

**Comment by gchatelet** [comment]
> It seems like all of these are public headers in `include`. I think we had those extra underscores for macro namespacing purposes. I'm not sure if we actually want to remove those, it might be better to remove the check from the include folder.

Or modify the clang tidy so it handles the `include` folder in a special way?
TBH I would prefer to have the clang tidy be the source of truth and have `clang-tidy --fix` do the job.

**Comment by nickdesaulniers** [comment]
> Is there a reason you used perl-as-sed rather than clang-tidy --fix?

It is one command line vs many different command line invocations. ;)

> I think that would fully canonicalize things wrt the trailing underscores too.

Sure, I'll give it a shot!

> It seems like all of these are public headers in include.
> I think we had those extra underscores for macro namespacing purposes.

Can you clarify what you mean by that?

> I'm not sure if we actually want to remove those, it might be better to remove the check from the include folder.
> Or modify the clang tidy so it handles the include folder in a special way?

Perhaps, if we have a justification why we need different conventions for different directories.  I highly suspect we don't actually need them and may be cargo culting historical accidents forward.

I could replace all these lines with `#pragma once` to further demonstrate the lack of a need for a convention based on directory (and remind users that we only support modern open source compilers from the last decade).

When in doubt, I guess I'd be curious if libcxx has such a convention between "user visible" vs "internal only" header guards?

**Comment by nickdesaulniers** [comment]
>> Is there a reason you used perl-as-sed rather than clang-tidy --fix?
>>I think that would fully canonicalize things wrt the trailing underscores too.

> Sure, I'll give it a shot!

So I hacked the ninja `libc-lint` target to just run `-checks="-*,llvm-header-guard" --fix` than ran `ninja libc-lint`.  While it cleaned up some headers, it also crashed a lot (ICE) and really mucked up some of the sources! I'm impressed!

**Comment by michaelrj-google** [comment]
By "macro namespacing" I mean ensuring all our publicly available macros start (and possibly end) with a specific prefix. Looking at it, it's probably fine for that prefix to just be `LLVM_LIBC_` since it's unlikely to have any collisions.

**Comment by nickdesaulniers** [comment]
Note: when `--fix` does work, it also renames some header guards. Example:
```diff
diff --git a/libc/src/__support/CPP/type_traits/is_fixed_point.h b/libc/src/__support/CPP/type_traits/is_fixed_point.h
index 317ba39748b7..025268bc2979 100644
--- a/libc/src/__support/CPP/type_traits/is_fixed_point.h
+++ b/libc/src/__support/CPP/type_traits/is_fixed_point.h
@@ -43,4 +43,4 @@ LIBC_INLINE_VAR constexpr bool is_fixed_point_v = is_fixed_point<T>::value;
 
 } // namespace LIBC_NAMESPACE::cpp
 
-#endif // LLVM_LIBC_SRC___SUPPORT_CPP_TYPE_TRAITS_IS_INTEGRAL_H
+#endif // LLVM_LIBC_SRC___SUPPORT_CPP_TYPE_TRAITS_IS_FIXED_POINT_H
diff --git a/libc/src/__support/GPU/generic/utils.h b/libc/src/__support/GPU/generic/utils.h
index 58db88dce1ca..3139604ac644 100644
```
(This is good, just heads up to reviewers that some guards will also be renamed in this PR).

>  it's probably fine for that prefix to just be LLVM_LIBC_ since it's unlikely to have any collisions.

:+1: SGTM

File: libc/include/llvm-libc-types/ACTION.h (hunk 1 above)
Comment (nickdesaulniers) [reply libc/include/llvm-libc-types/ACTION.h:9]: I haven't yet read the sources of this clang-tidy check, but a comment in a test for it seems to betray that I might need to manually fix up trailing underscores.

https://github.com/llvm-mirror/clang-tools-extra/blob/5c40544fa40bfb85ec888b6a03421b3905e4a4e7/unittests/clang-tidy/LLVMModuleTest.cpp#L60

**Comment by nickdesaulniers** [comment]
OK, I've figured a new way to automate applying fixits. I'ved added it to the commit message. PTAL

**Review by lntue — APPROVED** [review]

**Review by michaelrj-google — APPROVED** [review]

**Comment by nickdesaulniers** [comment]
Probably don't want header guards on assert.h: https://github.com/llvm/llvm-project/pull/83334

**Pull request merged**